### PR TITLE
[REVIEW] Databricks + RAPIDS + MLFlow + Hyperopt examples.

### PR DESCRIPTION
**Pull request by drobison00**

### CODE REVIEW

File: databricks/src/rf_test/train.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            df = cudf.read_orc(fpath)
+        else:
+            df = cudf.read_parquet(fpath)
+
+    return df
+
+
+def _train(params, fpath, mode='GPU', log_to_mlflow=None, hyperopt=False):
+    max_depth, max_features, n_estimators = params
+    max_depth, max_features, n_estimators = int(max_depth), float(max_features), int(n_estimators)
```
Comment (JohnZed) [inline databricks/src/rf_test/train.py:60]: how about combining these two lines?
`max_depth, max_features, n_estimators = (int(p) for p in params)`

File: databricks/src/rf_test/train.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    df = load_data(fpath, compute_type=mode)
+
+    for col in df.columns:
+        df[col] = df[col].astype('float32')
```
Comment (JohnZed) [inline databricks/src/rf_test/train.py:65]: I think these are already float32?

File: databricks/src/rf_test/train.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        mod = sk_RF(max_depth=max_depth, max_features=max_features, n_estimators=n_estimators)
+        acc_scorer = sk_acc
+    else:
+        raise RuntimeError("Unknown option. Choose between [CPU, GPU], case sensitive.")
+
+    mod.fit(X_train, y_train)
+    preds = mod.predict(X_test)
+    acc = acc_scorer(y_test, preds)
+
+    if (log_to_mlflow):
```
Comment (JohnZed) [inline databricks/src/rf_test/train.py:85]: maybe just always log to mlflow since that's core to the example and lightweight?

File: databricks/src/rf_test/train.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    if (not hyperopt):
+        return mod
+
+    return {'loss': acc, 'status': STATUS_OK}
+
+
+def train(params, fpath, mode='GPU', log_to_mlflow=None, hyperopt=False):
+    """
+    An example train method that computes the square of the input.
```
Comment (JohnZed) [inline databricks/src/rf_test/train.py:103]: Docstring needs updating.

File: databricks/src/rf_test/train.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                raise NotImplemented("CPU Remote read not implemented")
+        else:
+            if (fpath.startswith('/') or fpath.startswith('file:')):
+                df = pd.read_parquet(fpath)
+            else:
+                raise NotImplemented("CPU Remote read not implemented")
+
+
+    elif ('GPU' in compute_type):
+        if ('ORC' == data_type):
```
[the remark this answers is not in the document]
Comment (drobison00) [reply databricks/src/rf_test/train.py:50]: Will do.

File: databricks/src/rf_test/train.py (hunk 1 above)
Comment (drobison00) [reply databricks/src/rf_test/train.py:60]: max_features needs to be a float here, its a scaling coefficient (0, 1.0]

File: databricks/src/rf_test/train.py (hunk 2 above)
Comment (drobison00) [reply databricks/src/rf_test/train.py:65]: If I leave this out, the model will complain about certain features not being float32.

File: databricks/README.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
+## Databricks Notebooks with MLFlow, RAPIDS, and Hyperopt
+#### Currently, Databricks does not directly support modifying the conda environment, as a result, there are some incompatibilities with RAPIDS 0.14
+### Upload RAPIDS 0.13 Init Script
+1. Copy `src/rapids_install_cuml0.13_cuda10.0_ubuntu16.04.sh` onto your Databricks dbfs file system.
```
[the remark this answers is not in the document]
Comment (drobison00) [reply databricks/README.md:4]: Yes, this looks like another option. I'll get a working example.

File: databricks/src/rf_test/train.py (hunk 3 above)
Comment (drobison00) [reply databricks/src/rf_test/train.py:85]: This was to allow me to use the train function to just build a new model, and return it, without actually re-logging the metrics. I can remove it and pull the final model construction out.

File: databricks/notebooks/databricks_mlflow_rapids.ipynb [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+    "import sys\n",
+    "import threading\n",
+    "from queue import Queue, Empty\n",
+    "\n",
+    "from functools import partial\n",
+    "\n",
+    "import mlflow\n",
+    "import mlflow.sklearn\n",
+    "\n",
+    "from cuml.metrics.accuracy import accuracy_score as cuml_acc\n",
```
Comment (Nanthini10) [inline databricks/notebooks/databricks_mlflow_rapids.ipynb:20]: Just for the sake of consistency between the two notebooks, we can avoid aliases. Use `cuml.ensemble.RandomForestClassifier`, `cuml.metrics.accuracy.accuracy_score`, etc., just so the user is clear that it comes from cuml.

File: databricks/MLproject [hunk 8]
```diff
@@ -0,0 +1,11 @@
+name: cuML RF test
```
Comment (JohnZed) [inline databricks/MLproject:1]: Let's move to a subdir (mlflow_project) per discussion.
Actually, if this can all be done without databricks at all, we could move to a top-level dir `cloud-ml-examples/mlflow`

File: databricks/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnZed) [reply databricks/README.md:4]: Any thoughts on getting it to an s3 bucket? You should have access to the rapids data one now used in examples.

File: databricks/src/rf_test/train.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+def load_data(fpath, compute_type):
+    """
+    Simple helper function for loading data to be used by CPU/GPU models.
+
+    :param fpath: Path to the data to be ingested
+    :param compute_type: [CPU|GPU]
+    :return: DataFrame wrapping the data at [fpath]. Data will be in either a Pandas or RAPIDS (cuDF) DataFrame
+    """
+    if 'CPU' in compute_type:
```
Comment (JohnZed) [inline databricks/src/rf_test/train.py:27]: I thought we were going to make this path GPU-only? I think that would be a bit clearer and simplify the example.

File: databricks/src/rf_test/train.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    with mlflow.start_run():
+        mlflow.set_tag("mlflow.runName", "RAPIDS-Hyperopt-Databricks")
+        argmin = fmin(fn=fn,
+                      space=search_space,
+                      algo=algorithm,
+                      max_evals=2,
+                      trials=trials)
+
+        print("===========")
+        fn = partial(train, fpath=args.fpath, mode=args.mode, hyperopt=False)
```
Comment (JohnZed) [inline databricks/src/rf_test/train.py:135]: I think everything is very clean and easy to understand up until this point, then it's sort of unclear why we're rummaging around with the conda environment. Can we separate out the conda writing (probably optional?) from core mlflow stuff?

File: databricks/src/rf_test/train.py (hunk 9 above)
Comment (drobison00) [reply databricks/src/rf_test/train.py:27]: I'd just removed the CPU path from the notebook. I can take it out of the CLI workflow as well.

File: databricks/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnZed) [reply databricks/README.md:4]: Ok, sounds like we can defer this due to permissions issues.

File: mlflow_project/README.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+#### CLI Based Workflow
+1. Create a new conda environment.
+    1. `$ conda create -f envs/conda.yaml`
+1. Train the model
+    1. `$ cd mlflow\_project`
+    1. The example project is described in [MLProject](https://www.mlflow.org/docs/latest/projects.html) file.
+        1. This can be edited to allow additional command line variables, specify conda environments, and training
+        parameters (see link for additional information).
+    1. Publish to local tracking server
+        1. Here we instruct mlflow to run our training routine locally, and publish the result's to the local file system.
```
Comment (JohnZed) [inline mlflow_project/README.md:14]:
```suggestion
        1. Here we instruct mlflow to run our training routine locally, and publish the results to the local file system.
```

File: mlflow_project/README.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        1. ```shell script
+          $ mlflow run file:///$PWD -b local\
+                                   -P conda-env=$PWD/envs/conda.yaml\
+                                   -P fpath=$PWD/airline_100000.orc
```
Comment (JohnZed) [inline mlflow_project/README.md:18]: looks good, but do we provide airline_100000.orc? I included `airline_small.parquet` in the bucket.. maybe that was a bad name though

File: mlflow_project/README.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    1. Select the successful run and find the 'Full Path' element
+        1. ![Example 1](imgs/example.png)
+    1. Deploy your model
+        1. `$ mlflow models serve -m [PATH_TO_MODEL] -p 55755`
+
+1. Query the deployed model with test data `src/sample_server_query.sh` example script.
+    1. `$ bash src/sample_server_query.sh`
+
+
+#### CLI Based Workflow + Databricks
```
Comment (JohnZed) [inline mlflow_project/README.md:35]: Maybe expand a little "#### Integrating the CLI-Based Workflow with the DataBricks cloud"

File: mlflow_project/src/rf_test/train.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+def load_data(fpath):
+    """
+    Simple helper function for loading data to be used by GPU models.
+
+    :param fpath: Path to the data to be ingested
+    :return: DataFrame wrapping the data at [fpath]. Data will be in either a Pandas or RAPIDS (cuDF) DataFrame
+    """
+    import cudf
+
+    df = cudf.read_parquet(fpath)
```
Comment (JohnZed) [inline mlflow_project/src/rf_test/train.py:23]: Thank you for the simplification to just GPU! Maybe we should have load_data return X_train, X_test, y_train, y_test instead of data now that it's so simple?
(I suggest we defer this to a future PR though so we get the basics in)

File: mlflow_project/src/rf_test/train.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    :param params: hyperparameters. Its structure is consistent with how search space is defined. See below.
+    :param fpath: Path or URL for the training data used with the model.
+    :param hyperopt: Use hyperopt for hyperparameter search during training.
+    :return: dict with fields 'loss' (scalar loss) and 'status' (success/failure status of run)
+    """
+    with mlflow.start_run(nested=True):
+        return _train(params, fpath, hyperopt)
+
+
+if (__name__ in ("__main__",)):
```
Comment (JohnZed) [inline mlflow_project/src/rf_test/train.py:77]: No need for parens - looks less like the standard `__name__ == '__main__'` idiom this way

File: mlflow_project/README.md (hunk 12 above)
Comment (drobison00) [reply mlflow_project/README.md:18]: Good catch. Updating now.
